server/ServerFSM.cpp

```diff
@@ -521,207 +570,247 @@ sc::result MPLobby::react(const LobbyUpdate& msg) {
-    // or the galaxy setup.
-    
-    // check if galaxy setup data changed
-    has_important_changes = has_important_changes || (m_lobby_data->m_seed != incoming_lobby_data.m_seed) ||
-        (m_lobby_data->m_size != incoming_lobby_data.m_size) ||
-        (m_lobby_data->m_shape != incoming_lobby_data.m_shape) ||
-        (m_lobby_data->m_age != incoming_lobby_data.m_age) ||
-        (m_lobby_data->m_starlane_freq != incoming_lobby_data.m_starlane_freq) ||
-        (m_lobby_data->m_planet_density != incoming_lobby_data.m_planet_density) ||
-        (m_lobby_data->m_specials_freq != incoming_lobby_data.m_specials_freq) ||
-        (m_lobby_data->m_monster_freq != incoming_lobby_data.m_monster_freq) ||
-        (m_lobby_data->m_native_freq != incoming_lobby_data.m_native_freq) ||
-        (m_lobby_data->m_ai_aggr != incoming_lobby_data.m_ai_aggr) ||
-        (m_lobby_data->m_new_game != incoming_lobby_data.m_new_game);
-
-    if (player_setup_data_changed) {
-        if (m_lobby_data->m_players.size() != incoming_lobby_data.m_players.size()) {
-            has_important_changes = true; // drop ready at number of players changed
-        } else {
-            for (std::pair<int, PlayerSetupData>& i_player : m_lobby_data->m_players) {
-                if (i_player.first < 0) // ignore changes in AI.
-                    continue;
-                int player_id = i_player.first;
-                bool is_found_player = false;
-                for (std::pair<int, PlayerSetupData>& j_player : incoming_lobby_data.m_players) {
-                    if (player_id == j_player.first) {
-                        has_important_changes = has_important_changes || IsPlayerChanged(i_player.second, j_player.second);
-                        is_found_player = true;
-                        break;
-                    }
+    if (server.m_networking.PlayerIsHost(message.SendingPlayer())) {
+
+        DebugLogger() << "Get message from host.";
+
+        static int AI_count = 1;
+        const GG::Clr CLR_NONE = GG::Clr(0, 0, 0, 0);
+
+        // assign unique names / colours to any lobby entry that lacks them, or
+        // remove empire / colours from observers
+        for (std::pair<int, PlayerSetupData>& entry : incoming_lobby_data.m_players) {
+            PlayerSetupData& psd = entry.second;
+            if (psd.m_client_type == Networking::CLIENT_TYPE_HUMAN_OBSERVER ||
+                psd.m_client_type == Networking::CLIENT_TYPE_HUMAN_MODERATOR)
+            {
+                psd.m_empire_color = CLR_NONE;
+                // On OSX the following two lines must not be included.
+                // Clearing empire name and starting species name from
+                // PlayerSetupData causes a weird crash (bus error) deep
+                // in GG code on OSX in the Multiplayer Lobby when selecting
+                // Observer or Moderator as client type.
+#ifndef FREEORION_MACOSX
```

Comment:
2b48e33349db5a7d02ff48ea821bfa642c6e764e is a fix for a very weird crash I put in. I *think* I vaguely remember that the crash turned up again soon after I committed that fix, and ultimately turned out to be caused by something completely different (don't ask me what, it's too long ago).

Apparently I forgot to check if my fix could be safely undone now, but that should probably be done at some point.